# Building and Training a Convolutional Neural Network on the Fashion MNIST Dataset using Keras

## This Python script uses the Keras library to build, train, and evaluate a Convolutional Neural Network (CNN) for the Fashion MNIST dataset.

Here's a brief summary of each section:

1. Import necessary libraries and modules: The script imports necessary modules from Keras, including the Fashion MNIST dataset, the Sequential model, Dense and Flatten layers, Conv2D for convolutional layers, and a utility function for categorical conversion.

2. Load and preprocess data: The Fashion MNIST dataset is loaded and split into training and testing sets. The pixel values of the images are normalized to be between 0 and 1. The images are reshaped to have one channel, which is required for the CNN.

3. Create the model: A Sequential model is created and configured with one convolutional layer (Conv2D), a flattening layer (Flatten), and a dense layer. The activation function for the Conv2D layer is 'relu', and for the Dense layer, it's 'softmax'.

4. Compile the model: The model is compiled with the 'adam' optimizer, 'sparse_categorical_crossentropy' as the loss function, and 'accuracy' as the metric.

5. Train the model: The model is trained on the training data for 10 epochs with a batch size of 64.

6. Evaluate the model: The trained model is evaluated on the test data, and the test accuracy is printed out.


In [4]:
from keras.datasets import fashion_mnist
from keras.models import Sequential
from keras.layers import Dense, Flatten
from keras.layers.convolutional import Conv2D
from keras.utils import to_categorical

# Load data
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Normalize pixel values to be between 0 and 1
train_images, test_images = train_images / 255.0, test_images / 255.0

# Reshape the images to have one channel
train_images = train_images.reshape((train_images.shape[0], 28, 28, 1))
test_images = test_images.reshape((test_images.shape[0], 28, 28, 1))

# Create the model
model = Sequential()
model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(Flatten())
model.add(Dense(10, activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(train_images, train_labels, epochs=10, batch_size=64)

# Evaluate the model
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('Test accuracy:', test_acc)

Epoch 1/10
938/938 [==============================] - 17s 17ms/step - loss: 0.4316 - accuracy: 0.8484
Epoch 2/10
938/938 [==============================] - 17s 18ms/step - loss: 0.2995 - accuracy: 0.8942
Epoch 3/10
938/938 [==============================] - 16s 17ms/step - loss: 0.2634 - accuracy: 0.9059
Epoch 4/10
938/938 [==============================] - 18s 19ms/step - loss: 0.2396 - accuracy: 0.9150
Epoch 5/10
938/938 [==============================] - 18s 19ms/step - loss: 0.2195 - accuracy: 0.9205
Epoch 6/10
938/938 [==============================] - 16s 17ms/step - loss: 0.2033 - accuracy: 0.9276
Epoch 7/10
938/938 [==============================] - 16s 17ms/step - loss: 0.1888 - accuracy: 0.9328
Epoch 8/10
938/938 [==============================] - 16s 17ms/step - loss: 0.1770 - accuracy: 0.9365
Epoch 9/10
938/938 [==============================] - 16s 17ms/step - loss: 0.1647 - accuracy: 0.9411
Epoch 10/10
313/313 [==============================] - 2s 5ms/step - loss: 0.2886 

## This output represents the training and testing process of the Convolutional Neural Network (CNN) on the Fashion MNIST dataset.

1. Training: The model is trained for 10 epochs. For each epoch, the model goes through the entire training dataset. The 'loss' and 'accuracy' are calculated after each epoch. The 'loss' is a measure of the model's error, and 'accuracy' is the proportion of correct predictions. The loss decreases and accuracy increases with each epoch, indicating that the model is learning and improving its performance.

- Epoch 1: The model starts with a loss of 0.4316 and accuracy of 0.8484.
- Epoch 10: By the end of the 10th epoch, the model has a loss of 0.1532 and accuracy of 0.9457. This shows that the model has significantly improved its performance over the 10 epochs.

2. Testing: After training, the model is evaluated on the test dataset. The test loss and accuracy are calculated. The test loss is 0.2886 and the test accuracy is 0.9053. This means that the model correctly predicts the class of the images in the test set about 90.53% of the time.

The difference between training accuracy (94.57%) and testing accuracy (90.53%) suggests that the model might be slightly overfitting on the training data. Overfitting is when a model performs well on the training data but less well on unseen data (like the test data). This could be addressed by techniques like regularization, dropout, or gathering more data.